ocaml-lsp-server/src/fmt.ml

```diff
@@ -1,40 +1,44 @@
 open Import
 
-let read_to_end (in_chan : in_channel) : string =
-  let buf = Buffer.create 0 in
-  let chunk_size = 1024 in
-  let chunk = Bytes.create chunk_size in
-  let rec go pos =
-    let actual_len = input in_chan chunk pos chunk_size in
-    if actual_len > 0 then (
-      Buffer.add_subbytes buf chunk 0 actual_len;
-      go pos
-    )
-  in
-  go 0;
-  Buffer.contents buf
-
 type command_result =
   { stdout : string
   ; stderr : string
   ; status : Unix.process_status
   }
 
-let run_command command stdin_value args : command_result =
-  let command =
-    match args with
-    | [] -> command
-    | _ -> Printf.sprintf "%s %s" command (String.concat ~sep:" " args)
+let run_command scheduler thread bin stdin_value args : command_result Fiber.t =
+  let open Fiber.O in
+  let stdin_i, stdin_o = Unix.pipe ~cloexec:true () in
+  let stdout_i, stdout_o = Unix.pipe ~cloexec:true () in
+  let stderr_i, stderr_o = Unix.pipe ~cloexec:true () in
+  let pid =
+    let args = Array.of_list (bin :: args) in
+    Unix.create_process bin args stdin_i stdout_o stderr_o |> Stdune.Pid.of_int
+  in
+  Unix.close stdin_i;
```

Comment:
Shouldn't the file descriptors be closed only when the process exits?

Reply:
I don't think so, because we are our only closing the end of the pipe that we aren't intending to use.